### ISSUE TITLE
return-void not accepted in function getAcceptedIssuers()

### ISSUE THREAD

**Issue by jorgejams88** [shroudedcode/apk-mitm#22]
Hi, 

Thanks for the tool, I am trying to use it but I got the following error:

java.lang.VerifyError: Verifier rejected class *redacted_name*:java.security.cert.X509Certificate[] package.name.of.class$subclass.getAcceptedIssuers() failed to verify java.security.cert.X509Certificate return-void not expected.

Is there anything I can do?

**Comment by shroudedcode** [comment]
I have an idea of what could be causing this, but I can't test out a solution without the APK. Could you share what app you're trying to patch (preferably as an [APKPure](https://apkpure.com/) link if it's on there)? In case you don't want to share the app publicly, you can send it to me via email (see [my GitHub profile](https://github.com/shroudedcode)) or [Telegram](https://t.me/shroudedcode).

**Comment by jorgejams88** [comment]
Hi, thanks for your answer:

To see what's going on, I tried the --wait flag. Here's the smali code:

```
.method public getAcceptedIssuers()[Ljava/security/cert/X509Certificate;
    return-void
    .locals 1

    const/4 v0, 0x0

    new-array v0, v0, [Ljava/security/cert/X509Certificate;

    return-object v0
.end method
```
I changed it to the following (removing return-void) and it compiled, however the verification was still there:

```
.method public getAcceptedIssuers()[Ljava/security/cert/X509Certificate;
    .locals 1

    const/4 v0, 0x0

    new-array v0, v0, [Ljava/security/cert/X509Certificate;

    return-object v0
.end method
```

I was wondering if the way of returning null is different than return-void. Could that be it? I tried return 0x0 but it would not compile.

Let me know what you think.

**Comment by shroudedcode** [comment]
I've implemented this feature entirely based on [this guide by OWASP](https://github.com/OWASP/owasp-mstg/blob/master/Document/0x05c-Reverse-Engineering-and-Tampering.md#patching-example-disabling-certificate-pinning) which suggests inserting `return-void` at the beginning of all three `X509TrustManager` methods. Maybe that can cause issues with the `getAcceptedIssuers` method though since the expected return type is an array of `X509Certificate` instances.

I just found [this other guide](https://mobile-security.gitbook.io/mobile-security-testing-guide/android-testing-guide/0x05g-testing-network-communication#verifying-the-server-certificate) that suggests returning an empty `X509Certificate` array instead of `null` (which `return-void` does as far as I know). This is probably the solution you need, but their code is in Java. If you can translate it to Smali, you should be able to fix your problem. Please let me know if you do, so I can make `apk-mitm` insert the right line(s) by default!

**Comment by shroudedcode** [comment]
I've just taken a closer look at the Smali code of the `getAcceptedIssuers` method and realized what it does:

```smali
.method public getAcceptedIssuers()[Ljava/security/cert/X509Certificate;
    .locals 1

	# Initialize register `v0` to `0x0` (basically 0)
    const/4 v0, 0x0

	# Create an array with length `v0` (which is 0) and save it in `v0`
    new-array v0, v0, [Ljava/security/cert/X509Certificate;

	# Return `v0` (which is now an empty array)
    return-object v0
.end method
```

This is exactly what the code from [the guide I linked](https://mobile-security.gitbook.io/mobile-security-testing-guide/android-testing-guide/0x05g-testing-network-communication#verifying-the-server-certificate) is doing:

```java
public X509Certificate[] getAcceptedIssuers() {
    // Create an empty array and return it
    return new java.security.cert.X509Certificate[] {};
}
```

It seems like the APK you're trying to patch *already has* a `getAcceptedIssuers` method that accepts all certificates, but you said:

> I changed it to the following (removing return-void) and it compiled, however **the verification was still there**

What makes you think that the verification is still there? Have you tried running the app and got some errors?

**Comment by jorgejams88** [comment]
I was about to tell you that, the network validation was not there anymore. Any certificate was accepted, it was my mistake to think the validation was still there.

**Comment by shroudedcode** [comment]
> Any certificate was accepted, it was my mistake to think the validation was there.

Gotcha, that's not a big deal. :smile:

Anyway, there's still an `apk-mitm` bug here. From my understanding it should not add `return-void` to `getAcceptedIssuers` and instead insert something like this (since other apps don't have these lines):

```smali
    .locals 1
    const/4 v0, 0x0
    new-array v0, v0, [Ljava/security/cert/X509Certificate;
    return-object v0
```

That should fix the compilation issue you ran into and more reliably disable certificate pinning for others. I'll get to work on this soon and will let you know when it's fixed. Thanks for reporting the issue!

**Comment by jorgejams88** [comment]
Thank you!!! Have a great weekend.

**Comment by shroudedcode** [comment]
@jorgejams88 I just released [version 0.8.1](https://github.com/shroudedcode/apk-mitm/releases/tag/v0.8.1) with a fix. You should no longer run into the `return-void not accepted in function getAcceptedIssuers()` error now.
